hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -251,15 +257,52 @@ public static HoodieIndex createUserDefinedIndex(HoodieWriteConfig config) {
    */
   private static <R> HoodieData<HoodieRecord<R>> getExistingRecords(
       HoodieData<Pair<String, String>> partitionLocations, HoodieWriteConfig config, HoodieTable hoodieTable) {
-    final Option<String> instantTime = hoodieTable
-        .getMetaClient()
+    HoodieTableMetaClient metaClient = hoodieTable.getMetaClient();
+    final Option<String> instantTime = metaClient
         .getActiveTimeline() // we need to include all actions and completed
         .filterCompletedInstants()
         .lastInstant()
         .map(HoodieInstant::requestedTime);
-    return partitionLocations.flatMap(p
-        -> new HoodieMergedReadHandle(config, instantTime, hoodieTable, Pair.of(p.getKey(), p.getValue()))
-        .getMergedRecords().iterator());
+    ReaderContextFactory<R> readerContextFactory = hoodieTable.getContext().getReaderContextFactory(metaClient);
+    if (instantTime.isEmpty()) {
+      return hoodieTable.getContext().emptyHoodieData();
+    }
+    return partitionLocations.flatMap(p -> {
+      Option<FileSlice> fileSliceOption = Option.fromJavaOptional(hoodieTable
+          .getHoodieView()
+          .getLatestMergedFileSlicesBeforeOrOn(p.getLeft(), instantTime.get())
+          .filter(fileSlice -> fileSlice.getFileId().equals(p.getRight()))
```

Comment:
`partitionLocations` contains partition path and file ID pairs so `getLatestMergedFileSlicesBeforeOrOn` can be called on the same partition path multiple times (which is how it works on master).  Wondering if there is a better way to batch the requests.

Reply:
The result will be cached in the FSView so it should be fast. If we wanted to batch, there would be some sort of group by or a collect that can cause shuffling which may introduce more overhead.